Redis support

newtovaux opened the pull request:
This PR adds Redis caching support.

The Redis server is specified using REDIS_HOST and REDIS_PORT environment variables.

newtovaux committed:
Added scaffolding for Redis support

newtovaux committed:
Build plugin note

newtovaux committed:
Amended notification for server environment vars

newtovaux committed:
Updated docs

newtovaux committed:
Amended notification for server environment vars

newtovaux committed:
Adding core redis functions

newtovaux committed:
Fixing coding style

newtovaux commented:
Apologies for not squashing my commits prior to PR.

newtovaux committed:
Added missing reference

newtovaux committed:
Print the cached version

stklcode commented:
2.5 years later the same idea ... :grin:  I've deleted my fork in the meantime, because we postponed this feature in the past.
https://github.com/stklcode/cachify/commit/c09af684fc4077baf5af11dbd0e67d5fb1f3427d ...

stklcode commented on inc/class-cachify-redis.php:
```diff
@@ -0,0 +1,259 @@
+
+		/* Set options & connect */
+		try {
+			self::$_redis->connect( getenv( 'REDIS_HOST' ), intval( getenv( 'REDIS_PORT' ) ) );
```
I personally don't have these environment variables set on any of my web hosts with Redis support. Quite common in containerized environments or for bootstrapping scripts though.

Correct me, if I missed something, but I think environment variables are quite uncommon in the context of WP. We should not rely on environment variables as only configuration source here. To keep it consistent with the _memcached_ backend ([cachify-memcached.php lines 265-276](https://github.com/pluginkollektiv/cachify/blob/develop/inc/class-cachify-memcached.php#L265-L276)) I'd suggest a filter, i.e. `cachify_redis_servers`.

We also should make the port optional. If a unix socket is used for a local Redis node, the port is optional ([ref](https://github.com/phpredis/phpredis#connect-open)).

newtovaux replied:
Corrected in https://github.com/pluginkollektiv/cachify/pull/252/commits/37334725df3c089f1f03b520f414fe7f36ff35a7

stklcode commented on inc/class-cachify-redis.php:
```diff
@@ -0,0 +1,259 @@
+	/**
+	 * Redis-Object
+	 *
+	 * @since  2.0.7
```
I'd simply add a `@since 2.4`  (or whatever target version we'll merge this into) to the class and omit the copied annotations from the variables and methods.

newtovaux replied:
Corrected in https://github.com/pluginkollektiv/cachify/pull/252/commits/cc63cb784648be1dafd3f24d9e92869e89378887

stklcode commented on inc/class-cachify-redis.php:
```diff
@@ -0,0 +1,259 @@
+		// return ! empty( $option );
+		return class_exists( 'Redis' )
+			&& getenv( 'REDIS_HOST' )
+			&& getenv( 'REDIS_PORT' );
```
See comment on the connection logic.

newtovaux replied:
Corrected in https://github.com/pluginkollektiv/cachify/pull/252/commits/4367952085fe997b71fb293a93cbbcdb6fb46ed3

stklcode commented on inc/class-cachify-redis.php:
```diff
@@ -0,0 +1,259 @@
+	 * Redis-Object
+	 *
+	 * @since  2.0.7
+	 * @var    object
```
Actually it's `Redis`  oder `Redis|null` ...

newtovaux replied:
Corrected in https://github.com/pluginkollektiv/cachify/pull/252/commits/cc63cb784648be1dafd3f24d9e92869e89378887

stklcode commented on inc/class-cachify.php:
```diff
@@ -1652,6 +1673,12 @@ private static function _get_tabs( $options ) {
+			/* Setup tab for REDIS */
+			$tabs['setup'] = array(
+				'name' => __( 'Setup', 'cachify' ),
+				'page' => 'setup/cachify.redis.nginx.php',
```
If the setup instructions are webserver agnostic (they are for now), there should be no _.nginx_ suffix.

The "nginx" setup instructions are designed to show _nginx_ related setup, e.g. direct cache access from the reverse proxy like this: https://www.nginx.com/resources/wiki/modules/redis/

newtovaux replied:
Removed setup instructions for redis in https://github.com/pluginkollektiv/cachify/pull/252/commits/7afb46632f5d3d7de4ff54a456c5546e44717926

stklcode commented on inc/setup/cachify.redis.nginx.php:
```diff
@@ -0,0 +1,21 @@
+?>
+
+<h2><?php esc_html_e( 'Redis setup', 'cachify' ); ?></h2>
+<p><?php esc_html_e( 'Please ensure the following variables are added to your environment', 'cachify' ); ?></p>
```
"environment" is not self explaining here. Should be pointed out clearly that the system or process environment is meant here and not something like the _wp-config.php_ or webserver variables.

newtovaux replied:
Removed setup instructions for redis in https://github.com/pluginkollektiv/cachify/pull/252/commits/7afb46632f5d3d7de4ff54a456c5546e44717926

stklcode committed:
fix CS for print_cache()

stklcode committed:
fix CS for print_cache()

stklcode committed:
Merge branch 'pluginkollektiv/develop' into newtovaux/develop

stklcode committed:
fix indentations

newtovaux committed:
Corrected target version, as per @stklcode review

newtovaux committed:
Using filter, as per @stklcode review comment

newtovaux committed:
Removed env var, as per @stklcode review comment

newtovaux committed:
Removed setup instructions for redis

newtovaux commented:
@stklcode Many thanks for your review. I've hopefully addressed all the points you made. I've removed the setup instructions page, as it didn't feel necessary.

stklcode committed:
Merge branch 'pluginkollektiv/develop' into 'newtovaux/develop'

pfefferle commented:
@stklcode is there still "1 change requested"?

stklcode committed:
Merge branch 'pluginkollektiv/develop' into 'newtovaux/develop'

Zodiac1978 commented:
@stklcode is there still "1 change requested"?

stklcode commented:
Doesn't look like that's the case anymore. All my requested changes have been resolved in the meantime.

pfefferle commented:
@newtovaux can you maybe resolve the conflict, so that we can merge it?

stklcode committed:
Merge branch 'pluginkollektiv/develop' into 'newtovaux/develop'

stklcode committed:
remove empty line from Cachify_REDIS class (fix CS)

stklcode commented:
Resolved the (trivial) merge conflict and the code style issue from updated definition (just an empty line).

stklcode commented:
Merged ... finally :smile: 

Thanks again @newtovaux for your contribution. Let's see if we can get the version done and release this to the public.
